src/Nix/Diff/Types.hs

```diff
@@ -0,0 +1,318 @@
+   such as "Maps from `extraOutputs` must not have
+   intersecting keys" and so on.
+   If you want to test these invariants, you have to rewrite
+   instances manually.
+-}
+
+data Changed a = Changed { before :: a, now :: a }
+  deriving stock (Eq, Show, Functor, Foldable, Traversable, Generic)
+
+instance (Arbitrary a) => Arbitrary (Changed a) where
+  arbitrary = Changed <$> arbitrary <*> arbitrary
+  shrink = genericShrink
+
+instance ToJSON a => ToJSON (Changed a) where
+  toJSON = changedToJSON toJSON
+
+instance FromJSON a => FromJSON (Changed a) where
+  parseJSON = changedFromJSON parseJSON
+
+newtype TextDiff = TextDiff {unTextDiff :: [Patience.Item Text]}
+  deriving stock (Eq, Show)
+
+instance Arbitrary TextDiff where
+  arbitrary = TextDiff <$> listOf arbitraryItem
+
+instance ToJSON TextDiff where
+  toJSON = listValue itemToJSON . squashDiff . unTextDiff
+    where
+      -- TODO: think about longest difference spans,
+      -- instead of printing each line/word/character as one item.
+      -- Perhaps, this squashing should be in Diff module,
+      -- Because trees before and after squashing would be printed
+      -- in a different way by Render.HumanReadable.
+      --
+      -- Don't forget to update Arbitrary instance for that type.
+
+      -- squashDiff (Patience.Old a : Patience.Old b : xs) =
+      --   squashDiff (Patience.Old (a <> b) : xs)
+      -- squashDiff (Patience.New a : Patience.New b : xs) =
+      --   squashDiff (Patience.New (a <> b) : xs)
+      -- squashDiff (Patience.Both a _ : Patience.Both b _ : xs) =
+      --   let ab = a <> b in squashDiff (Patience.Both ab ab : xs)
+      -- squashDiff (x: xs) = x : squashDiff xs
+      -- squashDiff [] = []
+      squashDiff = id
+
+instance FromJSON TextDiff where
+  parseJSON v = TextDiff <$> (traverse itemFromJSON =<< parseJSON v)
+
+-- Helpfull aliases
+
+type OutputHash = Text
+
+type Platform = Text
+
+type Builder = Text
+
+type Argument = Text
+
+-- Derivation diff
+
+data DerivationDiff
+  = DerivationsAreTheSame
+  | AlreadyCompared
+  | NamesDontMatch   { outputStructure :: Changed OutputStructure}
+  | OutputsDontMatch { outputStructure :: Changed OutputStructure}
+  | DerivationDiff
+      { outputStructure :: Changed OutputStructure
+      , outputsDiff     :: OutputsDiff
+      , platformDiff    :: Maybe (Changed Platform)
+        -- ^ Will be Nothing, if Platform does not change
+      , builderDiff     :: Maybe (Changed Builder)
+        -- ^ Will be Nothing, if Builder does not change
+      , argumentsDiff   :: Maybe ArgumentsDiff
+        -- ^ Will be Nothing, if arguments are equal
+      , sourcesDiff     :: SourcesDiff
+      , inputsDiff      :: InputsDiff
+      , envDiff         :: Maybe EnvironmentDiff
+        -- ^ Will be Nothing, if environment comparison is skipped
+      }
+  deriving stock (Eq, Show, Generic)
+  deriving anyclass (ToJSON, FromJSON)
+  deriving Arbitrary via GenericArbitrary DerivationDiff
```

Comment:
When applying the first commit I receive the following error which is a bit to much for my basic haskel knowledge.

```
nix-diff> [1 of 3] Compiling Nix.Diff.Types   ( src/Nix/Diff/Types.hs, dist/build/Nix/Diff/Types.o, dist/build/Nix/Diff/Types.dyn_o )
nix-diff>
nix-diff> src/Nix/Diff/Types.hs:115:12: error:
nix-diff>     • solveWanteds: too many iterations (limit = 4)
nix-diff>         Unsolved: WC {}
nix-diff>         Set limit with -fconstraint-solver-iterations=n; n=0 for no limit
nix-diff>     • When deriving the instance for (Arbitrary DerivationDiff)
nix-diff>     |
nix-diff> 115 |   deriving Arbitrary via GenericArbitrary DerivationDiff
nix-diff>     |            ^^^^^^^^^
```

Reply:
Are you doing `nix-build` ?

Reply:
No, I fetched the first commit of this PR as a patch in my overlay and then added the missing dependencies. 

```nix
  haskellPackages = prev.haskellPackages.override (oldArgs: {
    overrides = lib.composeExtensions (oldArgs.overrides or (_: _: { }))
      (hfinal: hprev: {
        nix-diff = haskell.lib.doJailbreak (hprev.buildFromCabalSdist (hprev.nix-diff.overrideAttrs (oldAttrs: {
          patches = oldAttrs.patches or [ ] ++ [
            (fetchpatch {
              url = "https://github.com/Gabriella439/nix-diff/compare/904249022275f36ffd877de12b4a7fc4dae05af6..01d885640d9115c9263f4be11917cf7081b3f259.patch";
              excludes = [ ".github/workflows/ci.yaml" ];
              sha256 = "sha256-clKr4OwzYWQtbCPIqZfW42cMoIKNVTasbu9NzeBV55k=";
            })
          # using executableHaskellDepends does not work
          propagatedBuildInputs = oldAttrs.propagatedBuildInputs or [ ] ++ (with hprev; [
            aeson
            generic-arbitrary
            quickcheck-instances
            tasty-quickcheck
            tasty-silver
            typed-process
          ]);

          # takes multiple minutes and is single threaded
          doCheck = false;
        })));
      });
  });
```

Reply:
Can you reproduce it in current state? I changed commit order

Reply:
Yes, it is probably related to the libraries used in nixpkgs or how I apply the patch. I'll try the flake and see how that goes.

Reply:
Ok, please, make an issue, if you will found any bug